app/Pasteboard.m

```diff
@@ -0,0 +1,247 @@
+#include <string.h>
+#import <UIKit/UIKit.h>
+#include "fs/poll.h"
+#include "fs/dyndev.h"
+#include "kernel/errno.h"
+#include "debug.h"
+
+/**
+ * buffer is dynamically sized buffer of size buffer_cap
+ * All writes go to it, and buffer_len is length of data held in buffer
+ */
+
+// Prepare for fd separation
+#define fd_priv(fd) fd->clipboard
+typedef struct fd clip_fd;
+
+#define INITIAL_BUFFER_CAP 1024
+// 8MB: https://stackoverflow.com/a/3523175
+#define MAXIMAL_BUFFER_CAP 8*1024*1024
+
+// If pasteboard contents were changed since file was opened,
+// all read operations on in return error
+static int check_read_generation(clip_fd *fd) {
+    UIPasteboard *pb = UIPasteboard.generalPasteboard;
+
+    uint64_t pb_gen = (uint64_t) pb.changeCount;
+    uint64_t fd_gen = fd_priv(fd).generation;
+
+    if (fd_gen == 0 || fd->offset == 0) {
+        fd_priv(fd).generation = pb_gen;
+    } else if (fd_gen != pb_gen) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static const char *get_data(clip_fd *fd, size_t *len) {
+    if (fd_priv(fd).buffer != NULL) {
+        *len = fd_priv(fd).buffer_len;
+        return fd_priv(fd).buffer;
+    }
+
+    if (check_read_generation(fd) != 0) {
+        return NULL;
+    }
+
+    NSString *contents = UIPasteboard.generalPasteboard.string;
+    *len = contents.length;
+    return contents.UTF8String;
+}
+
+static int realloc_to_fit(clip_fd* fd, size_t fit_len) {
+    // (Re)allocate buffer if there's not enough space to fit fit_len
+    if (fit_len <= fd_priv(fd).buffer_cap) {
+        return 0;
+    }
+    if (fit_len > MAXIMAL_BUFFER_CAP) {
+        return 1;
+    }
+
+    size_t size = fd_priv(fd).buffer_cap * 2;
+    if (size == 0) {
+        size = INITIAL_BUFFER_CAP;
+    }
+    while (size < fit_len) size *= 2;
+
+    void *new_buf = realloc(fd_priv(fd).buffer, size);
+    if (new_buf == NULL) {
+        return 1;
+    }
+
+    fd_priv(fd).buffer = new_buf;
+    fd_priv(fd).buffer_cap = size;
+
+    return 0;
+}
+
+// buffer => UIPasteboard
+static int clipboard_wsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer == NULL) {
+        return 0;
+    }
+
+    void *data = fd_priv(fd).buffer;
+    size_t len = fd_priv(fd).buffer_len;
+
+    // FIXME(stek29): This logs "Returning local object of class NSString"
```

Comment:
Do you still get this warning on the latest version of the PR?

Reply:
Yup.